File: workers/api/src/activation-predicates.ts
import type { ProcessContract, Slot, SlotRule } from "./contracts";

export type SlotEvaluation = {
  predicate: string;
  expected: unknown;
  observed: unknown;
  passed: boolean;
  code: string;
};

export type PredicateContext = {
  contract: ProcessContract;
  receipt: Record<string, unknown>;
  slot: Slot;
  now?: Date;
};

const CONTENT_HASH = /^[0-9a-f]{64}$/;

function result(rule: SlotRule, observed: unknown, passed: boolean, code: string, expected: unknown): SlotEvaluation {
  return { predicate: rule.predicate, expected, observed, passed, code };
}

function present(value: unknown): boolean {
  return Boolean(String(value ?? "").trim());
}

function authorityExpected(context: PredicateContext): string[] {
  return context.contract.allowed_who?.length ? context.contract.allowed_who : ["authenticated_authority"];
}

function continuationExpected(rule: SlotRule, context: PredicateContext): string[] {
  if (rule.values?.length) return rule.values;
  const defaults: Record<"if_ok" | "if_doubt" | "if_not", string> = {
    if_ok: context.contract.process_id,
    if_doubt: String(context.contract.doubt_path ?? "attention-raise.v1"),
    if_not: "stop",
  };
  return [defaults[context.slot as "if_ok" | "if_doubt" | "if_not"]];
}

function parseTime(value: unknown): number | null {
  if (typeof value !== "string" || !/(?:Z|[+-]\d{2}:\d{2})$/.test(value)) return null;
  const parsed = Date.parse(value);
  return Number.isFinite(parsed) ? parsed : null;
}

type Predicate = (rule: SlotRule, value: unknown, context: PredicateContext) => SlotEvaluation;

const evaluatePresent: Predicate = (rule, value, context) => {
  const passed = present(value);
  return result(rule, value, passed, passed ? "present" : `${context.slot}_missing`, "non_empty");
};

const evaluateAuthorized: Predicate = (rule, value, context) => {
  const allowed = context.contract.allowed_who ?? [];
  const passed = present(value) && (!allowed.length || allowed.includes(String(value)));
  const failure = context.slot === "who" ? "who_not_authorized" : "confirmed_by_not_authorized";
  return result(rule, value, passed, passed ? "authorized" : failure, authorityExpected(context));
};

const evaluateAllowed: Predicate = (rule, value) => {
  const expected = rule.values ?? [];
  const passed = expected.length > 0 && expected.includes(String(value));
  return result(rule, value, passed, passed ? "did_admitted" : "did_not_admitted", expected);
};

const evaluateCanonical: Predicate = (rule, value) => {
  const passed = present(value);
  return result(rule, value, passed, passed ? "this_canonical" : "this_not_canonical", "non_empty_canonical_reference");
};

const evaluateHash: Predicate = (rule, value, context) => {
  const passed = CONTENT_HASH.test(String(value ?? ""));
  const failure = context.slot === "this" ? "this_not_content_hash" : "confirmation_evidence_invalid";
  return result(rule, value, passed, passed ? "content_hash" : failure, "64_lower_hex");
};

const evaluateRegisteredAt: Predicate = (rule, value) => {
  const passed = parseTime(value) !== null;
  return result(rule, value, passed, passed ? "when_registered_at" : "when_invalid", "timezone_aware_iso8601");
};

const evaluateFuture: Predicate = (rule, value, context) => {
  const now = context.now ?? new Date();
  const parsed = parseTime(value);
  const passed = parsed !== null && parsed > now.getTime();
  return result(rule, value, passed, passed ? "when_future" : "when_not_future", `after:${now.toISOString()}`);
};

const evaluateCompatible: Predicate = (rule, value, context) => {
  const expected = continuationExpected(rule, context);
  const passed = expected.includes(String(value));
  return result(
    rule,
    value,
    passed,
    passed ? `${context.slot}_compatible` : `${context.slot}_incompatible`,
    expected,
  );
};

const evaluateInitial: Predicate = (rule, value) => {
  const expected = rule.values?.length ? rule.values : ["registered"];
  const passed = expected.includes(String(value));
  return result(rule, value, passed, passed ? "status_initial" : "status_initial_invalid", expected);
};

const PREDICATES: Record<string, Predicate> = {
  "who.present": evaluatePresent,
  "who.authorized": evaluateAuthorized,
  "did.present": evaluatePresent,
  "did.allowed": evaluateAllowed,
  "this.present": evaluatePresent,
  "this.canonical": evaluateCanonical,
  "this.content_hash": evaluateHash,
  "when.registered_at": evaluateRegisteredAt,
  "when.future": evaluateFuture,
  "confirmed_by.present": evaluatePresent,
  "confirmed_by.authority": evaluateAuthorized,
  "confirmed_by.evidence_hash": evaluateHash,
  "if_ok.compatible": evaluateCompatible,
  "if_doubt.compatible": evaluateCompatible,
  "if_not.compatible": evaluateCompatible,
  "status.initial": evaluateInitial,
};

export function evaluateSlot(rule: SlotRule, value: unknown, context: PredicateContext): SlotEvaluation {
  const predicate = PREDICATES[rule.predicate];
  return predicate
    ? predicate(rule, value, context)
    : result(rule, value, false, "unknown_predicate", rule.predicate);
}

File: lab/activation_predicates.py
"""Closed predicate vocabulary for process activation."""

from __future__ import annotations

from datetime import datetime, timezone
import re
from typing import Any, Callable

from .contracts import ProcessContract, SlotRule


CONTENT_HASH = re.compile(r"^[0-9a-f]{64}$")


def _result(
    rule: SlotRule,
    observed: Any,
    *,
    passed: bool,
    code: str,
    expected: Any,
) -> dict[str, Any]:
    return {
        "predicate": rule.predicate,
        "expected": expected,
        "observed": observed,
        "passed": passed,
        "code": code,
    }


def _present(value: Any) -> bool:
    return bool(str(value or "").strip())


def _values(rule: SlotRule) -> list[str]:
    return list(rule.values)


def _parse_time(value: Any) -> datetime | None:
    try:
        parsed = datetime.fromisoformat(str(value).replace("Z", "+00:00"))
    except (TypeError, ValueError):
        return None
    if parsed.tzinfo is None:
        return None
    return parsed.astimezone(timezone.utc)


def _authority_expected(context: dict[str, Any]) -> list[str]:
    contract: ProcessContract = context["contract"]
    return list(contract.allowed_who) or ["authenticated_authority"]


def _continuation_expected(rule: SlotRule, context: dict[str, Any]) -> list[str]:
    if rule.values:
        return _values(rule)
    contract: ProcessContract = context["contract"]
    slot = context["slot"]
    defaults = {
        "if_ok": contract.process_id,
        "if_doubt": contract.doubt_path,
        "if_not": "stop",
    }
    return [defaults[slot]]


def _eval_present(rule: SlotRule, value: Any, context: dict[str, Any]) -> dict[str, Any]:
    slot = context["slot"]
    passed = _present(value)
    return _result(
        rule,
        value,
        passed=passed,
        code="present" if passed else f"{slot}_missing",
        expected="non_empty",
    )


def _eval_authorized(rule: SlotRule, value: Any, context: dict[str, Any]) -> dict[str, Any]:
    expected = _authority_expected(context)
    allowed = context["contract"].allowed_who
    passed = _present(value) and (not allowed or str(value) in allowed)
    slot = context["slot"]
    code = "authorized" if passed else (
        "who_not_authorized" if slot == "who" else "confirmed_by_not_authorized"
    )
    return _result(rule, value, passed=passed, code=code, expected=expected)


def _eval_allowed(rule: SlotRule, value: Any, context: dict[str, Any]) -> dict[str, Any]:
    expected = _values(rule)
    passed = bool(expected) and str(value) in expected
    return _result(
        rule,
        value,
        passed=passed,
        code="did_admitted" if passed else "did_not_admitted",
        expected=expected,
    )


def _eval_canonical(rule: SlotRule, value: Any, context: dict[str, Any]) -> dict[str, Any]:
    passed = _present(value)
    return _result(
        rule,
        value,
        passed=passed,
        code="this_canonical" if passed else "this_not_canonical",
        expected="non_empty_canonical_reference",
    )


def _eval_hash(rule: SlotRule, value: Any, context: dict[str, Any]) -> dict[str, Any]:
    passed = bool(CONTENT_HASH.fullmatch(str(value or "")))
    slot = context["slot"]
    code = "content_hash" if passed else (
        "this_not_content_hash" if slot == "this" else "confirmation_evidence_invalid"
    )
    return _result(rule, value, passed=passed, code=code, expected="64_lower_hex")


def _eval_registered_at(rule: SlotRule, value: Any, context: dict[str, Any]) -> dict[str, Any]:
    passed = _parse_time(value) is not None
    return _result(
        rule,
        value,
        passed=passed,
        code="when_registered_at" if passed else "when_invalid",
        expected="timezone_aware_iso8601",
    )


def _eval_future(rule: SlotRule, value: Any, context: dict[str, Any]) -> dict[str, Any]:
    parsed = _parse_time(value)
    now = context.get("now") or datetime.now(timezone.utc)
    passed = parsed is not None and parsed > now
    return _result(
        rule,
        value,
        passed=passed,
        code="when_future" if passed else "when_not_future",
        expected=f"after:{now.isoformat()}",
    )


def _eval_compatible(rule: SlotRule, value: Any, context: dict[str, Any]) -> dict[str, Any]:
    expected = _continuation_expected(rule, context)
    passed = str(value) in expected
    slot = context["slot"]
    return _result(
        rule,
        value,
        passed=passed,
        code=f"{slot}_compatible" if passed else f"{slot}_incompatible",
        expected=expected,
    )


def _eval_initial(rule: SlotRule, value: Any, context: dict[str, Any]) -> dict[str, Any]:
    expected = _values(rule) or ["registered"]
    passed = str(value) in expected
    return _result(
        rule,
        value,
        passed=passed,
        code="status_initial" if passed else "status_initial_invalid",
        expected=expected,
    )


Predicate = Callable[[SlotRule, Any, dict[str, Any]], dict[str, Any]]
PREDICATES: dict[str, Predicate] = {
    "who.present": _eval_present,
    "who.authorized": _eval_authorized,
    "did.present": _eval_present,
    "did.allowed": _eval_allowed,
    "this.present": _eval_present,
    "this.canonical": _eval_canonical,
    "this.content_hash": _eval_hash,
    "when.registered_at": _eval_registered_at,
    "when.future": _eval_future,
    "confirmed_by.present": _eval_present,
    "confirmed_by.authority": _eval_authorized,
    "confirmed_by.evidence_hash": _eval_hash,
    "if_ok.compatible": _eval_compatible,
    "if_doubt.compatible": _eval_compatible,
    "if_not.compatible": _eval_compatible,
    "status.initial": _eval_initial,
}


def evaluate_slot(
    rule: SlotRule,
    value: Any,
    context: dict[str, Any],
) -> dict[str, Any]:
    predicate = PREDICATES.get(rule.predicate)
    if predicate is None:
        return _result(
            rule,
            value,
            passed=False,
            code="unknown_predicate",
            expected=rule.predicate,
        )
    return predicate(rule, value, context)
